File: problem0004/c++/problem0004.cpp
/**
 * Problem: http://projecteuler.net/index.php?section=problems&id=4
 * Author: Elie De Brauwer <elie @ de-brauwer.be>
 * License: Simplified BSD
 * Date: December 24, 2010
 */

#include <algorithm>
#include <iostream>
#include <string>
#include <boost/lexical_cast.hpp>

/**
 * Check if a value is a palindrome.
 * @param val - The value which needs to be checked.
 * @return True when the value is a palindrome.
 */ 
bool isPalindrome(int val)
{
    std::string a = boost::lexical_cast<std::string>(val);
    std::string b = std::string(a);
    std::reverse(b.begin(),b.end());
    return a == b;
}

int main()
{
    int max = 0;
    for (int i=999; i>0; i--)
    {
        for (int j=999; j>0; j--)
        {
            int prod = i * j;
            if (prod > max && isPalindrome(prod))
            {
                max = prod;
            }
        }
    }
    std::cout << "Largest palindrome " << max << std::endl;
    return 0;
}

File: problem0004/java/Problem0004.java
/**
 * Problem: http://projecteuler.net/index.php?section=problems&id=4
 * Author: Elie De Brauwer <elie @ de-brauwer.be>
 * License: Simplified BSD
 * Date: December 24, 2010
 */

import java.lang.StringBuffer;

class Problem0004{

    /**
     * Check if a given value is a palindrome.
     * @param val The value which needs to be checked.
     * @return True when value is a palindrome.
     */
    public static boolean isPalindrome(int val){
	String normal = Integer.toString(val);
	String reverse = new StringBuffer(Integer.toString(val)).reverse().toString();
	return normal.compareTo(reverse) == 0;
    }

    public static void main(String [] args){
	int max = 0;
	for (int a = 999; a > 0; a--){
	    for (int b = 999; b >= a; b--){
		int prod = a*b;
		if (prod > max && isPalindrome(prod)){
		    max = prod;
		}
	    }
	}
	System.out.println("Largest palindrome is " + max);
    }
}
